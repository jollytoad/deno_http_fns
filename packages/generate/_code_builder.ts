import { Import } from "ts-poet";

export { type Code, code, joinCode, literalOf } from "ts-poet";

export function importAll(moduleSpec: string | URL, importedName: string) {
  return Import.importsAll(importedName, moduleSpec.toString());
}

export function importDefault(moduleSpec: string | URL, importedName: string) {
  return Import.importsDefault(importedName, moduleSpec.toString());
}

export function importNamed(
  moduleSpec: string | URL,
  exportedName: string,
  importedName?: string,
) {
  return Import.importsName(
    exportedName,
    moduleSpec.toString(),
    false,
    importedName,
  );
}
