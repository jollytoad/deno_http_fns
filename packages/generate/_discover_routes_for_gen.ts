import { discoverRoutes } from "@http/discovery/discover-routes";
import { combinedRouteMapper } from "@http/discovery/combined-route-mapper";
import type { GenerateOptions } from "./types.ts";
import type { DiscoveredRoute } from "@http/discovery/types";

export async function discoverRoutesForGen(
  opts: GenerateOptions,
): Promise<DiscoveredRoute[]> {
  const {
    pattern = "/",
    fileRootUrl,
    verbose = false,
  } = opts;

  const pathMapper = opts.pathMapper
    ? (await import(opts.pathMapper.toString())).default
    : undefined;

  const routeMapper = opts.routeMapper
    ? Array.isArray(opts.routeMapper)
      ? combinedRouteMapper(
        ...await Promise.all(opts.routeMapper.map(async (routeMapper) =>
          (await import(routeMapper.toString())).default
        )),
      )
      : (await import(opts.routeMapper.toString())).default
    : undefined;

  const compare = opts.compare
    ? (await import(opts.compare.toString())).default
    : undefined;

  return await discoverRoutes({
    pattern,
    fileRootUrl,
    pathMapper,
    routeMapper,
    compare,
    consolidate: true,
    verbose,
  });
}
