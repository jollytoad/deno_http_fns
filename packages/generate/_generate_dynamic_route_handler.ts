import type { DynamicRouteOptions } from "@http/discovery/dynamic-route";
import {
  type Code,
  code,
  importDefault,
  importNamed,
  joinCode,
  literalOf,
  relativeModulePath,
} from "./code_builder.ts";
import type {
  GenerateDynamicRouteOptions,
  HandlerGeneratorModule,
} from "./types.ts";

/**
 * Generate a route handler that uses `dynamicRoute` to discovered and
 * construct the filesystem router at runtime.
 */
export async function generateDynamicRouteHandler(
  opts: GenerateDynamicRouteOptions,
): Promise<Code> {
  const dynamicRoute = importNamed(
    `${opts.httpModulePrefix}discovery/dynamic-route`,
    "dynamicRoute",
  );

  const fileRootUrl = code`import.meta.resolve(${
    literalOf(relativeModulePath(opts.fileRootUrl, opts.moduleOutUrl))
  })`;

  const pattern = opts.pattern !== undefined
    ? literalOf(opts.pattern)
    : undefined;

  const pathMapper = opts.pathMapper
    ? importDefault(opts.pathMapper, "pathMapper")
    : undefined;

  const routeMapper = opts.routeMapper
    ? Array.isArray(opts.routeMapper)
      ? combineRouteMappers(opts, opts.routeMapper)
      : importDefault(opts.routeMapper, "routeMapper")
    : undefined;

  const handlerMapper = opts.handlerGenerator
    ? Array.isArray(opts.handlerGenerator)
      ? cascadeHandlerMappers(opts, await Promise.all(opts.handlerGenerator))
      : importDefault(
        (await opts.handlerGenerator).handlerMapper,
        "handlerMapper",
      )
    : undefined;

  const compare = opts.compare
    ? importDefault(opts.compare, "compare")
    : undefined;

  const { routeDiscovery: eagerness, verbose } = opts;

  const optsCode: Partial<Record<keyof DynamicRouteOptions, unknown>> = {
    ...(pattern ? { pattern } : null),
    fileRootUrl,
    eagerness,
    ...(pathMapper ? { pathMapper } : null),
    ...(routeMapper ? { routeMapper } : null),
    ...(handlerMapper ? { handlerMapper } : null),
    ...(compare ? { compare } : null),
    consolidate: true,
    ...(verbose ? { verbose } : null),
  };

  return code`${dynamicRoute}(${optsCode})`;
}

function cascadeHandlerMappers(
  opts: GenerateDynamicRouteOptions,
  handlerGenerators: Array<HandlerGeneratorModule>,
) {
  const cascadingHandlerMapper = importNamed(
    `${opts.httpModulePrefix}discovery/cascading-handler-mapper`,
    "cascadingHandlerMapper",
  );

  return code`${cascadingHandlerMapper}(${
    joinCode(
      handlerGenerators.map(({ handlerMapper }, i) =>
        code`${importDefault(handlerMapper, `handlerMapper_${i + 1}`)}`
      ),
      { on: "," },
    )
  })`;
}

function combineRouteMappers(
  opts: GenerateDynamicRouteOptions,
  routeMappers: Array<string | URL>,
) {
  const combinedRouteMapper = importNamed(
    `${opts.httpModulePrefix}discovery/combined-route-mapper`,
    "combinedRouteMapper",
  );

  return code`${combinedRouteMapper}(${
    joinCode(
      routeMappers.map((mapper, i) =>
        code`${importDefault(mapper, `routeMapper_${i + 1}`)}`
      ),
      { on: "," },
    )
  })`;
}
