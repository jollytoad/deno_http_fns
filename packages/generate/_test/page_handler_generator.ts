import type { Code, GeneratorOptions, RouteModule } from "@http/generate/types";
import {
  asCodePattern,
  asFn,
  awaitImportNamed,
  importNamed,
  returnFromFn,
} from "@http/generate/code-builder";
import { hasBodyFunction } from "$test/generate/page_handler_mapper.ts";

export const handlerMapper = "$test/generate/page_handler_mapper.ts";

/**
 * An example custom code generator for route modules that export a
 * function named `body`. The static code generation equivalent of
 * the [example handler mapper](./page_handler_mapper.ts).
 */
export function generate(
  { pattern, module, loaded }: RouteModule,
  { httpModulePrefix, moduleImports }: GeneratorOptions,
  i: number,
): Code | undefined {
  if (hasBodyFunction(loaded)) {
    const byPattern = asFn(importNamed(
      `${httpModulePrefix}route/by-pattern`,
      "byPattern",
    ));

    const pageHandler = asFn(importNamed(
      `$test/generate/page_handler.ts`,
      "pageHandler",
    ));

    const codePattern = asCodePattern(pattern);

    switch (moduleImports) {
      case "dynamic": {
        const lazy = asFn(importNamed(`${httpModulePrefix}route/lazy`, "lazy"));
        const body = awaitImportNamed(module, "body");

        return byPattern(codePattern, lazy(returnFromFn(pageHandler(body))));
      }

      case "static": {
        const routeModule = importNamed(
          module,
          "body",
          `page_body_${i}`,
        );

        return byPattern(codePattern, pageHandler(routeModule));
      }
    }
  }
}
