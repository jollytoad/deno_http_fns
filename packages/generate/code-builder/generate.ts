import type { Code, Import, Resolver } from "./types.ts";

export function generateModule(code: Code): string {
  return toImportLines(code).join("\n") + "\n\n" + code.toString();
}

export function resolveImports(
  code: { imports?: Import[] },
  resolver: Resolver,
) {
  code.imports?.forEach((imp) =>
    imp.moduleSpecifier = resolver(imp.moduleSpecifier)
  );
}

export function toImportLines({ imports }: { imports?: Import[] }): string[] {
  if (!imports) {
    return [];
  }

  const lines: string[] = [];

  const importsByExportKind = Object.groupBy(imports, exportKind);

  // importsByExportKind.none?.forEach(i => { lines.push(`import "${i.moduleSpecifier}";`) });
  importsByExportKind.all?.toSorted(byModuleSpecifier).forEach((i) => {
    lines.push(
      `import * as ${i.importedBinding} from "${i.moduleSpecifier}";`,
    );
  });
  importsByExportKind.default?.toSorted(byModuleSpecifier).forEach((i) => {
    lines.push(
      `import ${i.importedBinding} from "${i.moduleSpecifier}";`,
    );
  });

  const importsBySpecifier = Object.groupBy(
    importsByExportKind.named?.toSorted(byModuleSpecifier) ?? [],
    (i) => i.moduleSpecifier,
  );

  Object.entries(importsBySpecifier).map(([moduleSpecifier, imports]) => {
    const importSpecifiers = new Set(
      imports?.map((i) =>
        i.importedBinding !== i.moduleExportName
          ? `${i.moduleExportName} as ${i.importedBinding}`
          : i.importedBinding
      ),
    );
    if (importSpecifiers.size) {
      lines.push(
        `import { ${
          [...importSpecifiers].toSorted().join(", ")
        } } from "${moduleSpecifier}";`,
      );
    }
  });

  return lines;
}

function exportKind(
  imp: Import,
): "all" | "default" | "named" | "none" | "inline" {
  if (imp.inline) return "inline";
  switch (imp.moduleExportName) {
    case "*":
      return "all";
    case "default":
      return "default";
    default:
      return imp.moduleExportName ? "named" : "none";
  }
}

function byModuleSpecifier(a: Import, b: Import) {
  return a.moduleSpecifier.localeCompare(b.moduleSpecifier);
}
