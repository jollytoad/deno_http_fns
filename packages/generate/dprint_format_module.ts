import {
  createStreaming,
  type Formatter,
  type GlobalConfiguration,
} from "@dprint/formatter";

export function dprintFormatModule(config: GlobalConfiguration = {
  lineWidth: 80,
  indentWidth: 2,
}): (url: string | URL, content: string) => Promise<string> {
  let tsFormatter: Promise<Formatter>;

  return async function formatModule(
    filePath: string | URL,
    fileText: string,
  ): Promise<string> {
    tsFormatter ??= loadFormatter(config);

    return (await tsFormatter).formatText({
      filePath: filePath.toString(),
      fileText,
    });
  };
}

async function loadFormatter(config?: GlobalConfiguration) {
  const formatter = await createStreaming(
    // check https://plugins.dprint.dev/ for latest plugin versions
    fetch("https://plugins.dprint.dev/typescript-0.91.6.wasm"),
  );

  if (config) {
    formatter.setConfig(config, {});
  }

  return formatter;
}
