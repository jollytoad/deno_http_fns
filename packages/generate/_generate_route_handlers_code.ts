import { dirname } from "@std/path/posix/dirname";
import { relative } from "@std/path/posix/relative";
import { fromFileUrl } from "@std/path/posix/from-file-url";
import {
  asFn,
  type Code,
  dynamicImport,
  importNamed,
} from "./code-builder/mod.ts";
import type {
  GenerateOptions,
  GeneratorOptions,
  HandlerGeneratorModule,
  RouteModule,
} from "./types.ts";
import { discoverRoutesForGen } from "./_discover_routes_for_gen.ts";
import type { DiscoveredRoute } from "@http/discovery/types";

/**
 * Generate the handler(s) code for each discovered route.
 *
 * The code of each handler does not attempt to match the request URL for the route,
 * that must be added by the consumer of this function, allowing alternative strategies
 * for the actual routing (eg. a flat router using cascade/byPattern, or a tree router
 * using byPathTree/byPattern, or some other most advance/performant strategy).
 */
export async function generateRouteHandlersCode(
  opts: GenerateOptions,
): Promise<Map<DiscoveredRoute, Code[]>> {
  const {
    moduleOutUrl,
    httpModulePrefix = "@http/",
    moduleImports = "dynamic",
  } = opts;

  const outPath = dirname(fromFileUrl(moduleOutUrl));

  const routes = await discoverRoutesForGen(opts);

  const routesCode: Map<DiscoveredRoute, Code[]> = new Map();

  let i = 1;

  for (const route of routes) {
    const { pattern, module } = route;

    const routeCode: Code[] = [];

    let modulePath = relative(outPath, fromFileUrl(module));
    if (modulePath[0] !== ".") {
      modulePath = "./" + modulePath;
    }

    const loaded = await import(String(module));

    const handlerGenerators = await Promise.all(
      opts.handlerGenerator
        ? Array.isArray(opts.handlerGenerator)
          ? opts.handlerGenerator
          : [opts.handlerGenerator]
        : [
          import("./default_handler_generator.ts"),
          import("./methods_handler_generator.ts"),
        ],
    ) as Partial<HandlerGeneratorModule>[];

    const routeModule: RouteModule = { pattern, module, loaded };
    const generatorOpts: GeneratorOptions = {
      moduleOutUrl,
      httpModulePrefix,
      moduleImports,
    };

    const lazy = asFn(importNamed(`${httpModulePrefix}route/lazy`, "lazy"));

    for (const { generate } of handlerGenerators) {
      let code = generate?.(routeModule, generatorOpts, i);

      if (code) {
        if (moduleImports === "dynamic") {
          code.imports?.forEach(dynamicImport);
        }

        if (code.returnsPromise?.()) {
          code = lazy(returnFromFn(code));
        }

        routeCode.push(code);

        i++;
        continue;
      }
    }

    routesCode.set(route, routeCode);
  }

  return routesCode;
}

function returnFromFn(code: Code): Code {
  return {
    imports: code.imports,
    toString() {
      return (code.hasAwaits?.() ? "async " : "") + `() => ${code}`;
    },
  };
}
