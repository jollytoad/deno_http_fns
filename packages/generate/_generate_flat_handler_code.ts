import { asFn, type Code, importNamed, literal } from "./code-builder/mod.ts";
import type { GenerateOptions } from "./types.ts";
import { generateRouteHandlersCode } from "./_generate_route_handlers_code.ts";
import { asSerializablePattern } from "@http/discovery/as-serializable-pattern";

/**
 * Generate a route handler of static pre-built routes using
 * code generators prior to runtime.
 */
export default async function generateFlatHandlerCode(
  opts: GenerateOptions,
): Promise<Code> {
  const {
    httpModulePrefix = "@http/",
  } = opts;

  const cascade = asFn(importNamed(
    `${httpModulePrefix}route/cascade`,
    "cascade",
  ));

  const byPattern = asFn(importNamed(
    `${httpModulePrefix}route/by-pattern`,
    "byPattern",
  ));

  const routesCode = await generateRouteHandlersCode(opts);

  const byPatternsCode = routesCode.entries().flatMap(([route, codes]) => {
    const codePattern = literal(asSerializablePattern(route.pattern));
    return codes.map((code) => byPattern(codePattern, code));
  });

  return cascade(...byPatternsCode);
}
