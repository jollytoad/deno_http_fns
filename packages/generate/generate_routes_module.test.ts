import { withFallback } from "@http/route/with-fallback";
import {
  type GenerateOptions,
  generateRoutesModule,
} from "./generate_routes_module.ts";
import { assertSnapshot } from "@std/testing/snapshot";
import { getAvailablePort } from "@std/net/get-available-port";
import {
  assertStatus,
  STATUS_CODE,
  type StatusCode,
} from "@http/assert/status";
import { assertStringIncludes } from "@std/assert";
import { dprintFormatModule } from "./dprint_format_module.ts";

const fileRootUrl = import.meta.resolve("./_test/routes");

Deno.test("Generate with startup time discovery", async (t) => {
  const moduleOutUrl = import.meta.resolve(
    "./_test/routes_startup_discovery.ts",
  );

  await testGenerateRoutesModule(t, {
    moduleOutUrl,
    routeDiscovery: "startup",
  });

  await testRoutes(t, moduleOutUrl);
});

Deno.test("Generate with request time discovery", async (t) => {
  const moduleOutUrl = import.meta.resolve(
    "./_test/routes_request_discovery.ts",
  );

  await testGenerateRoutesModule(t, {
    moduleOutUrl,
    routeDiscovery: "request",
  });

  await testRoutes(t, moduleOutUrl);
});

Deno.test("Generate with static imports", async (t) => {
  const moduleOutUrl = import.meta.resolve("./_test/routes_static_imports.ts");

  await testGenerateRoutesModule(t, {
    moduleOutUrl,
    moduleImports: "static",
  });

  await testRoutes(t, moduleOutUrl);
});

Deno.test("Generate with dynamic imports", async (t) => {
  const moduleOutUrl = import.meta.resolve("./_test/routes_dynamic_imports.ts");

  await testGenerateRoutesModule(t, {
    moduleOutUrl,
    moduleImports: "dynamic",
  });

  await testRoutes(t, moduleOutUrl);
});

Deno.test("Generate with default opts", async (t) => {
  const moduleOutUrl = import.meta.resolve("./_test/routes_default_opts.ts");

  await generateRoutesModule({
    fileRootUrl,
    moduleOutUrl,
  });

  await assertSnapshot(t, await Deno.readTextFile(new URL(moduleOutUrl)));
});

async function testGenerateRoutesModule(
  t: Deno.TestContext,
  opts: Pick<
    GenerateOptions,
    "moduleOutUrl" | "moduleImports" | "routeDiscovery"
  >,
) {
  await t.step("generate module", async (t) => {
    await generateRoutesModule({
      ...opts,
      fileRootUrl,
      formatModule: dprintFormatModule(),
      routeMapper: [
        "$test/generate/ignored_route_mapper.ts",
        "@http/discovery/ts-route-mapper",
        "$test/generate/txt_route_mapper.ts",
      ],
      handlerGenerator: [
        import("$test/generate/page_handler_generator.ts"),
        import("@http/generate/default-handler-generator"),
        import("@http/generate/methods-handler-generator"),
      ],
    });

    await assertSnapshot(
      t,
      await Deno.readTextFile(new URL(opts.moduleOutUrl)),
    );
  });
}

async function testRoutes(t: Deno.TestContext, routesModule: string) {
  await t.step("test routes", async (t) => {
    // import the generated routes module
    const routes = (await import(routesModule)).default;

    // start a server using the routes
    const port = getAvailablePort();
    const ready = Promise.withResolvers();
    await using _server = Deno.serve({
      port,
      onListen: () => ready.resolve(`http://localhost:${port}`),
    }, withFallback(routes));
    const baseUrl = await ready.promise;

    // hit the server
    await testRoute(t, "GET", `${baseUrl}`, STATUS_CODE.OK);
    await testRoute(t, "GET", `${baseUrl}/about`, STATUS_CODE.OK);
    await testRoute(t, "GET", `${baseUrl}/unknown`, STATUS_CODE.NotFound);
    await testRoute(t, "GET", `${baseUrl}/user/bob`, STATUS_CODE.OK);
    await testRoute(t, "PUT", `${baseUrl}/user/bob`, STATUS_CODE.Accepted);
    await testRoute(
      t,
      "DELETE",
      `${baseUrl}/user/bob`,
      STATUS_CODE.MethodNotAllowed,
    );
    await testRoute(
      t,
      "GET",
      `${baseUrl}/raw`,
      STATUS_CODE.OK,
      async (response) => {
        const content = await response.text();
        assertStringIncludes(content, "Hello");
      },
    );
    await testRoute(
      t,
      "GET",
      `${baseUrl}/page`,
      STATUS_CODE.OK,
      async (response) => {
        const content = await response.text();
        assertStringIncludes(content, "<h1>Hello</h1>");
        assertStringIncludes(content, "<!DOCTYPE html>");
      },
    );
  });
}

async function testRoute(
  t: Deno.TestContext,
  method: string,
  url: string,
  expectedStatus: StatusCode,
  checkResponse?: (response: Response) => Promise<void>,
) {
  await t.step(`${method} ${url} -> ${expectedStatus}`, async () => {
    const response = await fetch(url, { method });
    assertStatus(response, expectedStatus);
    if (checkResponse) {
      await checkResponse(response);
    }
    if (response.body && !response.bodyUsed) {
      response.body.cancel();
    }
  });
}
