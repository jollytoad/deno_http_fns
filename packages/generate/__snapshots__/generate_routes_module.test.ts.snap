export const snapshot = {};

snapshot[`Generate with startup time discovery > generate module 1`] = `
'// IMPORTANT: This file has been automatically generated, DO NOT edit by hand.

import routeMapper_1 from "\$test/generate/ignored_route_mapper.ts";
import handlerMapper_1 from "\$test/generate/page_handler_mapper.ts";
import routeMapper_3 from "\$test/generate/txt_route_mapper.ts";
import { cascadingHandlerMapper } from "@http/discovery/cascading-handler-mapper";
import { combinedRouteMapper } from "@http/discovery/combined-route-mapper";
import handlerMapper_2 from "@http/discovery/default-handler-mapper";
import { dynamicRoute } from "@http/discovery/dynamic-route";
import handlerMapper_3 from "@http/discovery/methods-handler-mapper";
import routeMapper_2 from "@http/discovery/ts-route-mapper";

export default dynamicRoute({
  "fileRootUrl": import.meta.resolve("./routes"),
  "eagerness": "startup",
  "routeMapper": combinedRouteMapper(
    routeMapper_1,
    routeMapper_2,
    routeMapper_3,
  ),
  "handlerMapper": cascadingHandlerMapper(
    handlerMapper_1,
    handlerMapper_2,
    handlerMapper_3,
  ),
  "consolidate": true,
});
'
`;

snapshot[`Generate with request time discovery > generate module 1`] = `
'// IMPORTANT: This file has been automatically generated, DO NOT edit by hand.

import routeMapper_1 from "\$test/generate/ignored_route_mapper.ts";
import handlerMapper_1 from "\$test/generate/page_handler_mapper.ts";
import routeMapper_3 from "\$test/generate/txt_route_mapper.ts";
import { cascadingHandlerMapper } from "@http/discovery/cascading-handler-mapper";
import { combinedRouteMapper } from "@http/discovery/combined-route-mapper";
import handlerMapper_2 from "@http/discovery/default-handler-mapper";
import { dynamicRoute } from "@http/discovery/dynamic-route";
import handlerMapper_3 from "@http/discovery/methods-handler-mapper";
import routeMapper_2 from "@http/discovery/ts-route-mapper";

export default dynamicRoute({
  "fileRootUrl": import.meta.resolve("./routes"),
  "eagerness": "request",
  "routeMapper": combinedRouteMapper(
    routeMapper_1,
    routeMapper_2,
    routeMapper_3,
  ),
  "handlerMapper": cascadingHandlerMapper(
    handlerMapper_1,
    handlerMapper_2,
    handlerMapper_3,
  ),
  "consolidate": true,
});
'
`;

snapshot[`Generate flat router with static imports > generate module 1`] = `
'// IMPORTANT: This file has been automatically generated, DO NOT edit by hand.

import { pageHandler } from "\$test/generate/page_handler.ts";
import { byMethod } from "@http/route/by-method";
import { byPattern } from "@http/route/by-pattern";
import { cascade } from "@http/route/cascade";
import * as route_methods_2 from "./handle_txt.ts";
import * as route_methods_4 from "./routes/about.ts";
import route_5 from "./routes/index.ts";
import { body as page_body_3 } from "./routes/page.ts";
import * as route_methods_1 from "./routes/user/:name/index.ts";

export default cascade(
  byPattern("/user/:name{/}?", byMethod(route_methods_1)),
  byPattern("/raw", byMethod(route_methods_2)),
  byPattern(
    "/page",
    pageHandler(page_body_3, import.meta.resolve("./routes/page.ts")),
  ),
  byPattern("/about", byMethod(route_methods_4)),
  byPattern("/", route_5),
);
'
`;

snapshot[`Generate flat router with dynamic imports > generate module 1`] = `
'// IMPORTANT: This file has been automatically generated, DO NOT edit by hand.

import { pageHandler } from "\$test/generate/page_handler.ts";
import { byMethod } from "@http/route/by-method";
import { byPattern } from "@http/route/by-pattern";
import { cascade } from "@http/route/cascade";
import { lazy } from "@http/route/lazy";

export default cascade(
  byPattern(
    "/user/:name{/}?",
    lazy(async () => byMethod(await import("./routes/user/:name/index.ts"))),
  ),
  byPattern(
    "/raw",
    lazy(async () => byMethod(await import("./handle_txt.ts"))),
  ),
  byPattern(
    "/page",
    lazy(async () =>
      pageHandler(
        (await import("./routes/page.ts")).body,
        import.meta.resolve("./routes/page.ts"),
      )
    ),
  ),
  byPattern(
    "/about",
    lazy(async () => byMethod(await import("./routes/about.ts"))),
  ),
  byPattern("/", lazy(async () => (await import("./routes/index.ts")).default)),
);
'
`;

snapshot[`Generate tree router with static imports > generate module 1`] = `
'// IMPORTANT: This file has been automatically generated, DO NOT edit by hand.

import { pageHandler } from "\$test/generate/page_handler.ts";
import { byMethod } from "@http/route/by-method";
import { byPathTree } from "@http/route/by-path-tree";
import { byPattern } from "@http/route/by-pattern";
import { GUARD, LEAF, WILD } from "@http/route/path-tree/symbols";
import * as route_methods_2 from "./handle_txt.ts";
import * as route_methods_4 from "./routes/about.ts";
import route_5 from "./routes/index.ts";
import { body as page_body_3 } from "./routes/page.ts";
import * as route_methods_1 from "./routes/user/:name/index.ts";

export default byPathTree({
  "user": {
    [WILD]: {
      [GUARD]: [byPattern("/user/:name{/}?", byMethod(route_methods_1))],
    },
  },
  "raw": { [LEAF]: [byMethod(route_methods_2)] },
  "page": {
    [LEAF]: [pageHandler(page_body_3, import.meta.resolve("./routes/page.ts"))],
  },
  "about": { [LEAF]: [byMethod(route_methods_4)] },
  "": { [LEAF]: [route_5] },
});
'
`;

snapshot[`Generate tree router with dynamic imports > generate module 1`] = `
'// IMPORTANT: This file has been automatically generated, DO NOT edit by hand.

import { pageHandler } from "\$test/generate/page_handler.ts";
import { byMethod } from "@http/route/by-method";
import { byPathTree } from "@http/route/by-path-tree";
import { byPattern } from "@http/route/by-pattern";
import { lazy } from "@http/route/lazy";
import { GUARD, LEAF, WILD } from "@http/route/path-tree/symbols";

export default byPathTree({
  "user": {
    [WILD]: {
      [GUARD]: [byPattern(
        "/user/:name{/}?",
        lazy(async () =>
          byMethod(await import("./routes/user/:name/index.ts"))
        ),
      )],
    },
  },
  "raw": {
    [LEAF]: [lazy(async () => byMethod(await import("./handle_txt.ts")))],
  },
  "page": {
    [LEAF]: [lazy(async () =>
      pageHandler(
        (await import("./routes/page.ts")).body,
        import.meta.resolve("./routes/page.ts"),
      )
    )],
  },
  "about": {
    [LEAF]: [lazy(async () => byMethod(await import("./routes/about.ts")))],
  },
  "": {
    [LEAF]: [lazy(async () => (await import("./routes/index.ts")).default)],
  },
});
'
`;

snapshot[`Generate with default opts 1`] = `
'// IMPORTANT: This file has been automatically generated, DO NOT edit by hand.

import { byMethod } from "@http/route/by-method";
import { byPattern } from "@http/route/by-pattern";
import { cascade } from "@http/route/cascade";
import { lazy } from "@http/route/lazy";

export default cascade(byPattern("/user/:name{/}?", lazy(async () => byMethod(await import("./routes/user/:name/index.ts")))), byPattern("/about", lazy(async () => byMethod(await import("./routes/about.ts")))), byPattern("/", lazy(async () => (await import("./routes/index.ts")).default)));'
`;
